# 46.6 async/await

제너레이터로 비동기 처리를 하면 코드가 장황해지고 가독성도 나빠지는 단점이 있다.  
ES8에서 도입된 async/await는 제너레이터보다 간단하고 가독성이 좋게 비동기 처리를 동기 처리처럼 구현할 수 있다.

async/await는 프로미스 기반으로 동작한다.
하지만 프로미스의 후속 처리 메서드(then/catch/finally) 없이 동기 처리처럼 프로미스가 처리 결과를 반환할 수 있도록 구현할 수 있다.

```js
async function fetchTodo() {
  const BASE_URL = "https://jsonplaceholder.typicode.com/todos/1";

  const response = await fetch(BASE_URL);
  const todo = await response.json();
  console.log(todo);
  // {userId: 1, id: 1, title: 'delectus aut autem', completed: false}
}

fetchTodo();
```

## async 함수

await 키워드는 반드시 async 함수 내부에서 사용해야 한다.  
async 함수가 명시적으로 프로미스를 반환하지 않더라도 암묵적으로 반환값을 resolve하는 프로미스를 반환한다.

```js
// async 함수 선언문
async function foo(n) {
  return n;
}
foo(1).then((v) => console.log(v)); // 1

// async 함수 표현식
const bar = async function (n) {
  return n;
};
bar(2).then((v) => console.log(v)); // 2

// async 화살표 함수
const baz = async (n) => n;
baz(3).then((v) => console.log(v)); // 3

// async 메서드
const obj = {
  async foo(n) {
    return n;
  },
};
obj.foo(4).then((v) => console.log(v)); // 4

// async 클래스 메서드
class MyClass {
  async bar(n) {
    return n;
  }
}
const myClass = new MyClass();
myClass.bar(5).then((v) => console.log(v)); // 5
```

클래스의 constructor 메서드는 async 메서드가 될 수 없다.  
클래스의 constructor 메서드는 인스턴스를 반환해야 하지만 async 함수는 언제나 프로미스를 반환해야 한다.

```js
class MyClass {
    async constructor() {}
    // 문법 에러!
}

const myClass = new MyClass();
```

## await 키워드

await 키워드는 프로미스가 비동기 처리가 수행된 상태(settled)가 될 때까지 대기하다가 settled 상태가 되면 프로미스가 resolve한 처리 결과를 반환한다.  
await 키워드는 반드시 프로미스 앞에서 사용해야 한다.

```js
// fetch 함수가 수행한 HTTP 요청에 대한 서버의 응답이 도척해서 fetch 함수가 변환한 프로미스가 settled 상태가 될 때까지 1은 대기한다.
// 프로미스가 settled 상태가 되면 프로미스가 resolve한 처리 결과가 res 변수에 할당된다.
const getGithubUserName = async (id) => {
  const res = await fetch(`https://api.github.com/users/${id}`); // 1
  const { name } = await res.json(); // 2
  console.log(name);
};

getGithubUserName("Hyun-h"); // hyeona
```

모든 프로미스에 await를 사용하는 것은 주의한다.  
비동기 처리가 서로 연관이 있는지, 순차적으로 수행을 해서 순서가 보장되어야 하는지를 고려한 후 사용한다.

## 에러 처리

비동기 처리를 위한 콜백 패턴은 에러 처리가 곤란하다.  
에러는 호출자 방향으로 전파되는데 비동기 함수의 콜백 함수를 호출한 것은 비동기 함수가 아니기 때문에 try ... catch 문을 사용해 에러를 캐치할 수 없다.

반면 async/await 에서 에러 처리는 try ... catch 문을 사용할 수 있다.  
콜백 함수를 인수로 전달받는 비동기 함수와는 달리 프로미스를 반환하는 비동기 함수는 명시적으로 호출할 수 있기 때문에 호출자가 명확하다.

```js
const foo = async () => {
  try {
    const wrongUrl = "https://wrong.url";

    const response = await fetch(wrongUrl);
    const data = await response.json();
  } catch (err) {
    console.error(err); // TypeError!
  }
};

foo();
```

async 함수 내에서 catch 문을 사용해서 에러를 처리하지 않으면 async 함수는 발생한 에러를 reject하는 프로미스를 반환한다.  
따라서 async 함수를 호출하고 Promise.prototype.catch 후속 처리 메서드를 사용해 에러를 캐치할 수 있다.

```js
const foo = async () => {
  const wrongUrl = "https://wrong.url";

  const response = await fetch(wrongUrl);
  const data = await response.json();
  return data;
};

foo().then(console.log).catch(console.error); // TypeError!
```
